Replaced DESFIRE_TRANSCEIVE by a function (TODO n°3)

broth-itk opened the pull request:
Next try without mixing up my branches.

Replaced DESFIRE_TRANSCEIVE by a function, updating all commands it used.

smortex commented on libfreefare/mifare_desfire.c:
```diff
@@ -141,49 +145,89 @@ static ssize_t	 read_data(FreefareTag tag, uint8_t command, uint8_t file_no, off
+
+	if (!msg) {
+		errno = EINVAL;
+		return -1;
```
`return` in the macro allows us to exit from the function that use the macro.  With this refactoring, `return` only returns from the `desfire_transceive()` function, and the calling function has to check if `desfire_transceive()` was successfull or not in order to determine if it should continue or propagate the error.  Each time.

One solution to avoid a lot of repetition would be to use `setjmp(3)` / `longjmp(3)` to have some way to "raise" and "catch" errors, but it would adds clutter to any function calling `desfire_transceive()`.  Far from ideal from my point of view.

Maybe a lighter / simpler solution would be to wrap calls to `desfire_transceive()` in a macro that only does this check, e.g.

```c
#define DESFIRE_TRANSCEIVE(...) \
  do { \
    int n;
    if ((n = desfire_transceive(...)) < 0)
      return n;
  } while (0)
```

I am pretty sure modern compilers allow us to be sure that `desfire_transceive()` is never called without checking it's return value.  Found `-Wunused-result` after a quick search but I have not tested it.

Node: name macros that have side effects with UPPER CASED names: DESFIRE_TRANSCEIVE :wink:

broth-itk replied:
Thanks for your review!
I will check and fix this.

broth-itk replied:
BTW: Didn't we want to move away from using a macro?
Yes, there will be more lines and redundant code when checking always for the return value.
On the other hand doing so will allow more flexibility for handling errors.

I would vote for keeping desfire_transceive function and add return code checking in each function.
Already now it took me some time to understand the "hidden" variables created by the BUFFER macros. Keep the code readable :)

smortex replied:
I don't really have an opinion here.  I think a macro is acceptable if it is not longer than a few lines.  Anything is probably better than the current code :-)

All the BUFFER* macros are also quite disgusting, feel free to refactor this hardly if you feel brave enough!

smortex commented on libfreefare/mifare_desfire.c:
```diff
@@ -141,49 +145,89 @@ static ssize_t	 read_data(FreefareTag tag, uint8_t command, uint8_t file_no, off
+	  })
+#else
+	#define mifare_desfire_transceive desfire_transceive
+#endif
```
For the sake of simplicity, add the debug code directly to the `desfire_transceive()` function:

```c
static int
desfire_transceive(FreefareTag tag, const char *msg, size_t msg_len, char *res, size_t res_size, size_t *res_len)
{
#ifdef WITH_DEBUG
    DEBUG_FUNCTION();
#endif
    // more code
}
```

broth-itk replied:
Sorry, won't work, been there ;)
Doing so will show always "desfire_transceive" in the debug output.
The compiler does not support a macro showing the calling function.
IMHO the script is mandarory. And it works

smortex replied:
Ah!  Didn't checked what the macro does :unamused: Yeah, tracing things in C is a PITA.

smortex commented:
You have not authorized maintainers to push changes to your branch.

Can you configure with `./configure CFLAGS="-Wall -Wextra -Werror -Wno-unused-variable -Wno-unused-function"` and fix the problems detected (a cast missing, and a couple of `char *` that should by `uint8_t *`?

Thanks!

I also feel that when compiling `WITH_DEBUG`, `MIFARE_DESFIRE_TRANSCEIVE()` is broken.

broth-itk commented:
> You have not authorized maintainers to push changes to your branch.

I added you as collaborator - hope this solves the problem :) 

>Can you configure with `./configure CFLAGS="-Wall -Wextra -Werror -Wno-unused-variable -Wno->unused-function"` and fix the problems detected (a cast missing, and a couple of `char *` that should by >`uint8_t *`?

Done, fixed similar issues with other parts of the lib as well.
BTW: Enabling "-Werror" on configure will fail and tell that openssl headers were missing.

>I also feel that when compiling `WITH_DEBUG`, `MIFARE_DESFIRE_TRANSCEIVE()` is broken.

"./configure --enable-debug" works just fine for me, no problems whatsoever

broth-itk commented:
Travis CI failes for missing nfc-utils header file. I did not touch that. Any idea what possibly can cause that problem?

broth-itk commented:
Is there anything left for me to do in order to get this pull request passing all checks and get eventually merged? Sorry for asking possibly dumb questions, the git process is still new to me. Thanks!

AdamLaurie commented:
test$ ./run-test.sh 
make: *** No rule to make target 'echo-cutter'. Stop.
./run-test.sh: 17: ./run-test.sh: : Permission denied

don't have time to dig into this but if someone has a quick fix let me know and i'll try again...

smortex commented:
@AdamLaurie if I recall correctly, this is the typical failure when cutter is not installed when the autotools where run.

Double-check that [cutter](http://cutter.sourceforge.net/) is installed, and re-run `autoreconf -is`, then `./configure` and `make check` with a formatted cards on NFC devices connected to the system.

Thanks!

darconeous commented:
I'll give testing this a shot tonight.

smortex commented:
Ah, indentation looks weird (maybe we should add a test for this in Travis-CI?).

I think you should be able to approve other people Pull Requests and merge them, but maybe if you contributed some code you cannot anymore?

Would you mind checking `make style`? Thanks!
